### DISCUSSION TITLE
Enhance test-related classes as loaded

### DISCUSSION THREAD

**Discussion by sebersole** [junit-team/junit-framework#3203] [Q&A]
Is there an existing extension point in JUnit 5 that would allow me to hook into the loading of test related classes, Java agent style?

The classes to be enhanced are specified by extension annotations. E.g.

```
@DomainModel( annotatedClasses = {SomeClass.class, AnotherClass.class} )
```

Here, I'd want `SomeClass` and `AnotherClass` to be enhanced as they get loaded.

**Comment by portyu9** [comment]
Hi @sebersole 

No, JUnit Jupiter does not have an extension point equivalent to a Java agent's class-load transformation.

JUnit extensions can participate in discovery/test lifecycle and invocation, but they do not receive class bytes **before the JVM defines an arbitrary class**.

For true load-time enhancement, use the Java instrumentation API:

```java
public static void premain(String args, Instrumentation instrumentation) {
    instrumentation.addTransformer(new MyClassFileTransformer());
}
```

with the agent supplied using `-javaagent`.

A `ClassFileTransformer` is invoked before the class definition is completed, which is the appropriate level for bytecode enhancement.

There is also an important detail with an annotation such as:

```java
@DomainModel(annotatedClasses = {SomeClass.class, AnotherClass.class})
```

If you discover those values from a JUnit extension, you are already dealing with `Class<?>` objects. That is too late if the requirement is to transform those classes before their initial definition.

If the agent needs to discover enhancement targets without loading them first, consider representing the targets as class names/metadata that the agent can match independently, rather than relying on a Jupiter extension to obtain the `Class<?>` values.

So the responsibilities should be separated:

- **JUnit extension:** test lifecycle/configuration
- **Java agent / `ClassFileTransformer`:** load-time bytecode enhancement

Please mark this answer as accepted if it helped, thank you.
